packages/common/solana/utils.ts

```diff
@@ -136,17 +135,97 @@ export async function signAndExecuteTransaction(
   hash: BlockhashWithExpiryBlockHeight,
 ): Promise<string> {
   const signedTx = await signTransaction(invoker, tx);
-  const rawTx = signedTx.serialize();
 
-  if (!hash.lastValidBlockHeight || !signedTx.signature || !hash.blockhash)
-    throw Error("Error with transaction parameters.");
+  return executeTransaction(connection, signedTx, hash);
+}
+
+/**
+ * Sends and confirms Transaction
+ * Confirmation strategy is not 100% reliable here as in times of congestion there can be a case that tx is executed,
+ * but is not in `commitment` state and so it's not considered executed by the `sendAndConfirmRawTransaction` method,
+ * and it raises an expiry error even though transaction may be executed soon.
+ * - so we add additional 50 blocks for checks to account for such issues;
+ * - also, we check for SignatureStatus one last time as it could be that websocket was slow to respond.
+ * @param connection - Solana client connection
+ * @param tx - Transaction instance
+ * @param hash - blockhash information, the same hash should be used in the Transaction
+ * @returns Transaction signature
+ */
+export async function executeTransaction(
+  connection: Connection,
+  tx: Transaction,
+  hash: BlockhashWithExpiryBlockHeight,
+): Promise<string> {
+  const rawTx = tx.serialize();
+
+  if (!hash.lastValidBlockHeight || !tx.signature || !hash.blockhash) throw Error("Error with transaction parameters.");
 
+  const signature = bs58.encode(tx.signature);
   const confirmationStrategy: BlockheightBasedTransactionConfirmationStrategy = {
     lastValidBlockHeight: hash.lastValidBlockHeight + 50,
-    signature: bs58.encode(signedTx.signature),
+    signature,
```

Comment:
Is anything from this article useful/applicable?
https://twitter.com/jordaaash/status/1774892864847368230